### DISCUSSION TITLE
Using EDS with cFS

### DISCUSSION THREAD

**Discussion by ugo94490** [nasa/cFS#284] [Q&A] [answered]
Hello, I am currently studying the functioning of EDS on cFS.

I started by exploring the cFS repository and I could find several xml describing each service of cFE.

I don't really understand how they are used, for example is there a tool that would allow to generate .h from the xml describing these EDS ?

Thanks to this post on the github discussions of cFS : https://github.com/nasa/cFS/discussions/274#discussioncomment-922995 , I could see that there were other repositories in the NASA Github organization that mentioned EDS.

Are these repositories fully integrated into cFS? If so what are their roles and where are they located in cFS ?

-Ugo

**Comment by jphickey** [comment, marked as answer]
EDS is a work in progress.  In the recent versions of CFE (since 6.6 IIRC), some EDS XML files describing core services are included in the source tree, however they are for reference only - they are not actually used by the build in the official GSFC releases at this time.

Separate projects have developed tools to utilize these files into a more complete workflow, for instance:

- to generate C structures (.h files) that match the binary interface
- to account for architecture specific differences (endianness, padding, etc) and provide a consistent CMD/TLM binary interface
- to provide bindings to scripting languages like Python and Lua
- to provide a functional test automation framework
- to facilitate generating a ground system database (e.g. Cosmos txt files and/or XTCE)

There are proof-of-concepts around for all of this stuff, but it still needs some agreement around the specific details of how the cFS interfaces should be described, the toolchain implementation(s) to settle on, the build artifacts, etc.

In short, its been a slow process (years in the making) but progress is gradually being made, and there seems to be some renewed community interest in getting something in place in a future release.

FWIW, a project I previously worked on released the tool and cFS integration under Apache, and can be found at https://github.com/nasa/edslib and https://github.com/jphickey/cfe-eds-framework (the latter just being a combined single repo containing all sub-modules that correlates with the cFE 6.8-rc2 baseline).

**Reply by ThibFrgsGmz** [reply]
Your cfe-eds-framework project is a good resource to learn how to handle EDS SOIS in general.

Just out of curiosity, on the SANA website https://sanaregistry.org/r/sois/ there is a SEDS (ccsds.sois.seds.xml) defining a collection of several types that can be found in base_types.xml of your project, didn't you use this file to define type names close to "stdin.h" or was this file not yet published when you started to develop the project ?

**Reply by jphickey** [reply]
Yes, these files were developed somewhat in parallel.  Ongoing work of the SOIS-WG should eventually consolidate this overlap/duplication.

**Reply by ivanperez-keera** [reply]
It's been over two years since this message. @jphickey Any updates? What is the status today?

**Comment by dmccomas** [comment]
Here's another cFS EDS reference/resource.  I have a project called [cFS Basecamp](https://github.com/cfs-tools/cfs-basecamp) described in this short [cFS basecamp Overview](https://youtu.be/jwV3_9W8dcY) video that uses an 'App Store' approach so new cFS targets can easily be assembled. The assembly process is made possible because of EDS command and telemetry definitions coupled with app JSON initialization parameter files.  I've recently started to publish [cFS-based educational projects](https://openmissionstack.com/projects) that rely on the app store.

From a broader cFS EDS perspective I think a tool that converts EDS to XTCE would be beneficial so there's a standards-based path to ground systems that support XTCE.
